### PR TITLE
feat: cache api version on auth info

### PR DESCRIPTION
**Pull request by amphro**
Currently, every connection instantiation will call out and get a list of API versions. This is a waste of time considering Salesforce servers only change API versions every few months. We want to pick up quickly when the server is updated so this sets a 24 hour cache. The call is quick so this seems appropriate. 

It also introduces the environment variable `SFDX_IGNORE_API_VERSION_CACHE` to disable this cache to return to old behavior.

This also fixes an old bug auth info files where the username is an access token write to disk.

Setting `DEBUG=*connection*` you will see something like the following:

No Cache:
```
sfdx:connection DEBUG Using the latest because lastChecked=undefined and SFDX_IGNORE_API_VERSION_CACHE=false +0ms
sfdx:connection DEBUG request: {"method":"GET","url":"https://adcxgs0hub.my.salesforce.com/services/data","headers":{"content-type":"application/json","user-agent":"sfdx toolbelt:"}} +157ms
sfdx:connection DEBUG response for org versions: 20.0,21.0,22.0,23.0,24.0,25.0,26.0,27.0,28.0,29.0,30.0,31.0,32.0,33.0,34.0,35.0,36.0,37.0,38.0,39.0,40.0,41.0,42.0,43.0,44.0,45.0,46.0,47.0,48.0,49.0,50.0,51.0 +395ms
sfdx:connection DEBUG Loaded latest apiVersion 51.0 +1ms
sfdx:connection DEBUG Using apiVersion 51.0 +1ms
```

Cached:
```
sfdx:connection DEBUG Last checked on 1613683877813 (now is 1613684745568) - not getting latest +0ms
sfdx:connection DEBUG Loaded latest apiVersion 51.0 +1ms
sfdx:connection DEBUG Using apiVersion 51.0 +0ms
```

Invalid Cache:
```
sfdx:connection DEBUG Last checked on 161683877813 (now is 1613684784405) - getting latest +0ms
sfdx:connection DEBUG request: {"method":"GET","url":"https://adcxgs0hub.my.salesforce.com/services/data","headers":{"content-type":"application/json","user-agent":"sfdx toolbelt:"}} +83ms
sfdx:connection DEBUG response for org versions: 20.0,21.0,22.0,23.0,24.0,25.0,26.0,27.0,28.0,29.0,30.0,31.0,32.0,33.0,34.0,35.0,36.0,37.0,38.0,39.0,40.0,41.0,42.0,43.0,44.0,45.0,46.0,47.0,48.0,49.0,50.0,51.0 +344ms
sfdx:connection DEBUG Loaded latest apiVersion 51.0 +2ms
sfdx:connection DEBUG Using apiVersion 51.0 +0ms
```

Disable Cache:
```
sfdx:connection DEBUG Using the latest because lastChecked=1613684784833 and SFDX_IGNORE_API_VERSION_CACHE=true +0ms
sfdx:connection DEBUG request: {"method":"GET","url":"https://adcxgs0hub.my.salesforce.com/services/data","headers":{"content-type":"application/json","user-agent":"sfdx toolbelt:"}} +298ms
sfdx:connection DEBUG response for org versions: 20.0,21.0,22.0,23.0,24.0,25.0,26.0,27.0,28.0,29.0,30.0,31.0,32.0,33.0,34.0,35.0,36.0,37.0,38.0,39.0,40.0,41.0,42.0,43.0,44.0,45.0,46.0,47.0,48.0,49.0,50.0,51.0 +361ms
sfdx:connection DEBUG Loaded latest apiVersion 51.0 +2ms
sfdx:connection DEBUG Using apiVersion 51.0 +0ms
```

@W-8856053@

### CODE REVIEW

File: src/connection.ts [hunk 1]
```diff
@@ -361,6 +390,44 @@ export class Connection extends JSForceConnection {
...
+      this.options.authInfo.save({
+        instanceApiVersion: version,
+        instanceApiVersionLastRetrieved: Date.now(),
+      });
+    };
+
+    const ignoreCache = env.getBoolean('SFDX_IGNORE_API_VERSION_CACHE', false);
+    if (lastChecked && !ignoreCache) {
+      const now = Date.now();
+      const has24HoursPastSinceLastCheck = now - lastChecked > Duration.hours(24).milliseconds;
```
Comment (peternhale) [inline src/connection.ts:413]: Would be reasonable to establish an env var or config setting to allow customer to control the TTL, with a default of 24 hours?

Comment (amphro) [reply src/connection.ts:413]: I'm open if you think that will be useful. I don't feel like it will be that useful. It is a quick call.

**Review by WillieRuemmele — APPROVED** [review]

File: src/authInfo.ts [hunk 2]
```diff
@@ -556,6 +558,12 @@ export class AuthInfo extends AsyncCreatable<AuthInfo.Options> {
   public async save(authData?: AuthFields): Promise<AuthInfo> {
     this.update(authData);
     const username = ensure(this.getUsername());
+
+    if (sfdc.matchesAccessToken(username)) {
+      this.logger.debug('Username is an accesstoken. Skip saving authinfo to disk.');
+      return this;
+    }
```
Comment (shetzel) [inline src/authInfo.ts:565]: 👍

File: src/connection.ts [hunk 3]
```diff
@@ -361,6 +390,44 @@ export class Connection extends JSForceConnection {
     }
     return result.records[0];
   }
+
+  private async loadInstanceApiVersion(): Promise<Optional<string | null>> {
```
Comment (shetzel) [inline src/connection.ts:394]: I think you want this instead:
`Promise<Nullable<string>>`

**Review by shetzel — APPROVED** [review]

File: src/authInfo.ts [hunk 4]
```diff
@@ -58,6 +58,8 @@ export interface AuthFields {
   createdOrgInstance?: string;
   devHubUsername?: string;
   instanceUrl?: string;
+  instanceApiVersion?: string;
+  instanceApiVersionLastRetrieved?: number;
```
Comment (mshanemc) [inline src/authInfo.ts:62]: `created` is a string (js date).  Pros:
* consistency
* human-readable in a file

Cons: extra step to parse?

**Review by mshanemc — APPROVED** [review]

File: src/authInfo.ts (hunk 4 above)
Comment (amphro) [reply src/authInfo.ts:62]: I can do that.

File: src/connection.ts (hunk 1 above)
Comment (amphro) [reply src/connection.ts]: I'll leave this for another item if we decide we want to offer the option.

**Pull request merged**